File: src/nodes/decorators/InverterNode.ts
import { TreeNode } from '../../core/TreeNode.js';
import { NodeStatus } from '../../core/NodeStatus.js';
import { Blackboard } from '../../core/Blackboard.js';
import { Theme } from '../../utils/Theme.js';

/**
 * InverterNode: Inverts the result of its child.
 *
 * - SUCCESS → FAILURE
 * - FAILURE → SUCCESS
 * - RUNNING → RUNNING
 *
 * Use case: "Do the opposite of X"
 */
export class InverterNode extends TreeNode {
    constructor() {
        super('inverter', 'Inverter', 'decorator', 'fa-exchange-alt', Theme.node.decorator);
        this.maxChildren = 1; // Decorators can only have one child
    }

    public tick(blackboard: Blackboard): NodeStatus {
        if (this.children.length === 0) {
            this.status = NodeStatus.FAILURE;
            return this.status;
        }

        // Tick the only child (decorators have single child)
        const child = this.children[0];
        const childStatus = child.tick(blackboard);

        // Invert success and failure
        if (childStatus === NodeStatus.SUCCESS) {
            this.status = NodeStatus.FAILURE;
        } else if (childStatus === NodeStatus.FAILURE) {
            this.status = NodeStatus.SUCCESS;
        } else {
            // RUNNING or IDLE stays the same
            this.status = childStatus;
        }

        return this.status;
    }
}

File: UnitySDK/Runtime/Nodes/InverterNode.cs
namespace WebBehaviorTree.Nodes
{
    /// <summary>
    /// Inverts the result of its child.
    /// Success becomes Failure, Failure becomes Success.
    /// Running remains Running.
    /// </summary>
    public class InverterNode : TreeNode
    {
        protected override NodeStatus OnTick(Blackboard blackboard)
        {
            if (Children.Count == 0)
            {
                return NodeStatus.Failure;
            }

            var childStatus = Children[0].Tick(blackboard);

            return childStatus switch
            {
                NodeStatus.Success => NodeStatus.Failure,
                NodeStatus.Failure => NodeStatus.Success,
                _ => childStatus
            };
        }
    }
}
